ddtrace/tracer/rules_sampler.go

```diff
@@ -0,0 +1,572 @@
+		}
+		return rules, nil
+	}
+	if rulesFile != "" {
+		rulesFromEnvFile, err := os.ReadFile(rulesFile)
+		if err != nil {
+			errs = append(errs, fmt.Sprintf("Couldn't read file from DD_SPAN_SAMPLING_RULES_FILE: %v", err))
+		}
+		spanRules, err = unmarshalSamplingRules(rulesFromEnvFile, SamplingRuleSpan)
+		if err != nil {
+			errs = append(errs, err.Error())
+		}
+		rules = append(rules, spanRules...)
+	}
+	if len(errs) != 0 {
+		return rules, fmt.Errorf("%s", strings.Join(errs, "\n\t"))
+	}
+	return rules, nil
+}
+
+// unmarshalSamplingRules unmarshals JSON from b and returns the sampling rules found, attributing
+// the type t to them. If any errors are occurred, they are returned.
+func unmarshalSamplingRules(b []byte, spanType SamplingRuleType) ([]SamplingRule, error) {
+	if len(b) == 0 {
+		return nil, nil
+	}
+	var jsonRules []struct {
+		Service      string      `json:"service"`
+		Name         string      `json:"name"`
+		Rate         json.Number `json:"sample_rate"`
+		MaxPerSecond float64     `json:"max_per_second"`
+	}
+	err := json.Unmarshal(b, &jsonRules)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshalling JSON: %v", err)
+	}
+	rules := make([]SamplingRule, 0, len(jsonRules))
+	var errs []string
+	for i, v := range jsonRules {
+		if v.Rate == "" {
+			errs = append(errs, fmt.Sprintf("at index %d: rate not provided", i))
+			continue
+		}
+		rate, err := v.Rate.Float64()
+		if err != nil {
+			errs = append(errs, fmt.Sprintf("at index %d: %v", i, err))
+			continue
+		}
+		if !(rate >= 0.0 && rate <= 1.0) {
+			log.Warn("at index %d: ignoring rule %+v: rate is out of [0.0, 1.0] range", i, v)
+			continue
+		}
+		switch spanType {
+		case SamplingRuleSpan:
+			if v.Service == "" {
+				v.Service = "*"
+			}
+			srvGlob, err := globMatch(v.Service)
+			if err != nil {
+				log.Warn("at index %d: ignoring rule %+v: service name regex pattern can't be compiled", i, v)
+				continue
+			}
+			if v.Name == "" {
+				v.Name = "*"
+			}
+			opGlob, err := globMatch(v.Name)
+			if err != nil {
+				log.Warn("at index %d: ignoring rule %+v: operation name regex pattern can't be compiled", i, v)
+				continue
+			}
+			rules = append(rules, SamplingRule{
+				Service:      srvGlob,
+				Name:         opGlob,
+				Rate:         rate,
+				MaxPerSecond: v.MaxPerSecond,
+				limiter:      newSingleSpanRateLimiter(v.MaxPerSecond),
+				Type:         SamplingRuleSpan,
+			})
+		case SamplingRuleTrace:
+			switch {
+			case v.Service != "" && v.Name != "":
+				rules = append(rules, NameServiceRule(v.Name, v.Service, rate))
+			case v.Service != "":
+				rules = append(rules, ServiceRule(v.Service, rate))
+			case v.Name != "":
+				rules = append(rules, NameRule(v.Name, rate))
+			}
+		}
+	}
+	if len(errs) != 0 {
+		return rules, fmt.Errorf("\n\t%s", strings.Join(errs, "\n\t"))
+	}
+	return rules, nil
+}
+
+// MarshalJSON implements the json.Marshaler interface.
+func (sr *SamplingRule) MarshalJSON() ([]byte, error) {
```

Comment:
Where are the tests for this (and other code in rules_sampler.go)?
side note: this function would be a great candidate for fuzzing

Reply:
I would have to ask whoever wrote that code 😅 Good point, I can add this either within this PR or open a new one to keep the scope aligned with the RFC